### ISSUE TITLE
About the Prototype-based Masked Cross-Attention

### ISSUE THREAD

**Issue by zzzyzh** [NiccoloCavagnero/PEM#11]
First of all, I want to express my appreciation for the excellent work you've done. I apologize for my limited coding skills, but I have a few questions regarding the implementation of the **Prototype-based Masked Cross-Attention module**:

1. Could you please point me to where this module is implemented in your code? It would be greatly helpful if you could provide a direct link to it.

2. How is the prototype updated in your implementation? Wouldn't using only an NxD prototype limit the precision of the prompts? Have you considered using an `NxCxD` prototype, where `C` represents the number of prototypes per class?

Thank you for your time and assistance!

**Comment by NiccoloCavagnero** [comment]
Thank you for your consideration!

1. You can find our efficient cross-attention in pem/modeling/transformer_decoder/pem_transformer_decoder.py

2. We do not store prototypes, so we do not update them, but we simply compute them on the fly taking, for each query, its most similar feature token. Ideally, if instances are well separated in feature spaces, a single token per instance should be sufficient, also considering the fact that all tokens of a given instance should be similar. You can for sure try to increase the number of selected tokens per query but I do believe that this won't help.

**Comment by zzzyzh** [comment]
Thank you very much for your patient response. I still have a small question: Why is `nn.Embedding` used to represent the prototype? What is the difference between using `nn.Embedding` and `nn.Parameter` for this purpose?

**Comment by NiccoloCavagnero** [comment]
You are welcome! 

Queries are defined as learnable parameters following Mask2Former, there exists better ways to initialise the queries in the detection literature, although these have not been tested yet in the context of segmentation.

With respect to the actual differences between nn.Embedding and nn.Parameter I would point you to official PyTorch documentation.
